library/core/src/sync/atomic.rs

```diff
@@ -243,7 +243,7 @@ const EMULATE_ATOMIC_BOOL: bool =
 
 /// A boolean type which can be safely shared between threads.
 ///
-/// This type has the same in-memory representation as a [`bool`].
+/// This type has the same size, alignment, and bit validity as a [`bool`].
```

Comment:
> Hmm, it's a bit odd to see "alignment" here when it's not in the others.

Ideally AtomicI8&AtomicU8 docs should mention that as well, but I don't know how easy it would be to do that due to macro.

> One thing I'm unsure about: this clearly has a safety invariant that it's zero or one, but is that strictly a validity invariant? How do those interact with atomic ops? Or is the API restricted enough that it never hits a scenario where the question would come up?

In my understanding, the stabilization of from_ptr (done in 1.75) establishes a requirement that they must have strictly the same validity invariant as bool.

https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicBool.html#method.from_ptr

Reply:
> Hmm, it's a bit odd to see "alignment" here when it's not in the others.
> 
> I guess it's a consequence of the other things, though: `bool` is size 1 per https://doc.rust-lang.org/std/mem/fn.size_of.html, which means the alignment can't be more than 1, and alignment is strictly positive so it's squeezed to have only one possible value.

I could also just explicitly document that the alignment is 1 if you'd prefer.

> One thing I'm unsure about: this clearly has a safety invariant that it's zero or one, but is that strictly a _validity_ invariant? How do those interact with atomic ops? Or is the API restricted enough that it never hits a scenario where the question would come up?

Does @taiki-e's argument convince you? If not, I'd be happy to omit this for `AtomicBool` in particular (which is the only type here with a non-trivial bit validity requirement). Alternatively, I could say something like "it's guaranteed that all bit-valid `bool`s can be soundly transmuted to `AtomicBool`, but the inverse is not currently guaranteed."

Reply:
> Alternatively, I could say something like "it's guaranteed that all bit-valid `bool`s can be soundly transmuted to `AtomicBool`, but the inverse is not currently guaranteed."

I only mentioned from_ptr(bool -> AtomicBool) because as_ptr(AtomicBool -> bool) was already stable at the time, but validity here should be guaranteed in both directions.

https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicBool.html#method.as_ptr

Reply:
Yeah, I agree. I just meant that the proposed wording could work as an alternative if @scottmcm isn't convinced by your argument that we can rely on bit validity being equal between the two types. In particular, it sounds like their concern is about whether atomic operations could introduce the possibility of an `AtomicBool` temporarily being in some in-between state. Unless I'm misunderstanding, that concern does not imply that 0x00 or 0x01 would be *invalid* `AtomicBool` instances, but rather that there might be *more* valid `AtomicBool` instances than just those two.

Reply:
> whether atomic operations could introduce the possibility of an AtomicBool temporarily being in some in-between state.

Hmm, if an atomic operation has an observable in-between state, it is not "atomic", right?

Reply:
Yeah, I agree; I'm just trying to steel-man @scottmcm's argument. I am convinced by your argument that bit validity is completely equivalent, but I want to make sure to leave the door open for other language changes on this PR if they aren't convinced.